### PR TITLE
adds support for asynchronous reads using the ReadListener from servlet api

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- adds support for asynchronous reads using the ReadListener from servlet api

## Why are we making these changes?

We would like to avoid manually managing a custom thread pool to pump the request bytes to the backend.

### CODE REVIEW

File: waiter/src/waiter/process_request.clj [hunk 1]
```diff
@@ -190,143 +190,107 @@
...
+     Returns the bytes read from the input stream."
+    [streaming-timeout-ms error-handler-fn ^ServletInputStream input-stream body-ch]
+    (let [bytes-available (.available input-stream)
+          buffer-size (-> bytes-available
+                        (bit-and buffer-increment-mask)
+                        (max min-buffer-size)
+                        (min max-buffer-size))
+          buffer-bytes (byte-array buffer-size)
+          bytes-read (.read input-stream buffer-bytes)]
+      (log/info "processed bytes from the request input stream" {:available bytes-available :read bytes-read})
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Should we switch this to `log/debug`?

File: waiter/src/waiter/process_request.clj [hunk 2]
```diff
@@ -190,143 +190,107 @@
...
+      (.setReadListener
+        input-stream
+        (reify ReadListener
+          (onAllDataRead [_]
+            (complete-request-streaming nil))
+          (onDataAvailable [_]
+            (try
+              (cid/with-correlation-id
+                correlation-id
+                (log/info "reading available bytes from request input stream")
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: Should we switch this to `log/debug`?

File: waiter/src/waiter/process_request.clj [hunk 3]
```diff
@@ -190,143 +190,107 @@
...
+                (loop []
+                  (let [bytes-read (read-available-bytes!! streaming-timeout-ms complete-request-streaming input-stream body-ch)]
+                    (when (pos? bytes-read)
+                      (report-request-size-metrics bytes-read false))
+                    (when (and (utils/non-neg? bytes-read) (.isReady input-stream))
```
Comment (sradack) [inline waiter/src/waiter/process_request.clj]: All the examples I've seen have `isReady()` guarding `read()` calls.  Should we check `isReady()` before calling `read()`?

Comment (shamsimam) [reply waiter/src/waiter/process_request.clj]: The example at https://docs.oracle.com/javaee/7/tutorial/servlets013.htm is using a `do ... while()` loops where the call to `read()` is before the call to `isReady()`

**Comment by shamsimam** [comment]
> What wasn't working before?

1. Using the ReadListener inside jet at https://github.com/twosigma/jet/blob/waiter-patch/src/clj/qbits/jet/servlet.clj#L71 was not working.

2. My previous attempts (before the thread pool pump) did not use the call to `isReady()` and were not receiving the `onDataAvailable()` callbacks.

3. I also received [`IllegalStateException: ReadListener already set` errors](https://github.com/twosigma/waiter/pull/804) before, but did not dig deeper since my `onDataAvailable()` callbacks were not working.

File: waiter/src/waiter/process_request.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/process_request.clj]: Done.

File: waiter/src/waiter/process_request.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/process_request.clj]: Done.

**Pull request merged**